plugins/Validation/src/Validation.ts

```diff
@@ -1,223 +1,346 @@
+/* tslint:disable:no-any */
+import { State } from '@hookstate/core';
+import { PluginCallbacks, StateValueAtRoot } from '@hookstate/core/dist';
 
-import { Plugin, Path, StateValueAtPath, State } from '@hookstate/core';
+export const ValidationId = Symbol('Validation');
 
-export type ValidationSeverity = 'error' | 'warning';
+type ValidateFn<T> = (value: T) => boolean;
+type Path = readonly (string | number | symbol)[];
 
-export interface ValidationError {
-    readonly message: string;
-    readonly path: Path;
-    readonly severity: ValidationSeverity;
+interface NestedState {
+    state: State<any>;
+    parent?: NestedState;
 }
 
-export interface ValidationExtensions<S> {
-    validate(attachRule: (value: S) => boolean,
-        message: string | ((value: S) => string),
-        severity?: ValidationSeverity): void,
-    validShallow(): boolean,
-    valid(): boolean,
-    invalidShallow(): boolean,
-    invalid(): boolean,
-    firstError(
-        filter?: (e: ValidationError) => boolean,
-        depth?: number
-    ): Partial<ValidationError>,
-    errors(
-        filter?: (e: ValidationError) => boolean,
-        depth?: number,
-        first?: boolean
-    ): ReadonlyArray<ValidationError>,
+interface SingleValidator<T> {
+    required(message?: string): void;
+
+    when(fn: (value: State<T>) => boolean): this;
+
+    validate(validator: ValidateFn<T>, message?: string): void;
 }
 
-const PluginID = Symbol('Validate');
+type NestedValidator<Root, T> = T extends string ? SingleValidator<T> :
+    T extends any[] ? ArrayValidator<Root, T[0]> : ObjectValidator<Root, T>;
 
-const emptyErrors: ValidationError[] = []
+type ObjectNestedValidator<Root, T> = {
+    [Key in keyof T]: NestedValidator<Root, T[Key]>
+};
 
-interface ValidationRule {
-    readonly message: string | ((value: StateValueAtPath) => string)
-    readonly rule: (v: StateValueAtPath) => boolean
-    readonly severity: ValidationSeverity;
+interface ObjectRootValidator<Root, T> extends SingleValidator<T> {
+    when(fn: (value: State<T>, root: Root) => boolean): this;
 }
 
-class ValidationPluginInstance<S> {
-    private storeRules = {};
+type ObjectValidator<Root, T> = ObjectNestedValidator<Root, T> & ObjectRootValidator<Root, T> & {
+    whenType<K extends keyof T, V extends T[K]>(key: K, value: V):
+        ObjectValidator<Root, T & { [Key in keyof Pick<T, K>]: V }>;
+};
 
-    getRulesAndNested(path: Path): [ValidationRule[], string[]] {
-        let result = this.storeRules;
-        path.forEach(p => {
-            if (typeof p === 'number') {
-                p = '*' // limitation: support only validation for each element of array
-            }
-            result = result && (result[p])
-        });
-        return [result && result[PluginID] ? Array.from(result[PluginID].values()) : [],
-            result ? Object.keys(result) : []];
+type ArrayValidator<Root, T> = ObjectValidator<Root, T> & {
```

Comment:
If Root is a type of root state object, I would make it as a second argument in generic parameters. And probably provided a default `unknown`. I actually though about capturing root state type in the core package and maybe make Plugin type with a generic Root state type parameter. But maybe in the future.

Reply:
i actually don't need to carry Root around anymore, but agree, would be nice for Plugin to by typed Plugin<T>